File: AIClient_JAVA/AIClient_JAVA/src/network/NetworkController.java
/* ------------------------------------------------------------------------------
 ** _________   _________      ________    _____      _____  ___________ _________
 ** \_   ___ \ /   _____/     /  _____/   /  _  \    /     \ \_   _____//   _____/
 ** /    \  \/ \_____  \     /   \  ___  /  /_\  \  /  \ /  \ |    __)_ \_____  \ 
 ** \     \____/        \    \    \_\  \/    |    \/    Y    \|        \/        \
 **  \______  /_______  /     \______  /\____|__  /\____|__  /_______  /_______  /
 **         \/        \/             \/         \/         \/        \/        \/ 
 **
 ** NetworkController.java
 ** Control all the network communication
 **
 ** Author: Samuel-Ricardo Carriere
 ** ------------------------------------------------------------------------------*/

package network;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.net.Socket;

import event.EventFactory;

public class NetworkController {
	private static NetworkController instance = null;
	private Socket webSocket;
	private PrintWriter outToServer;
	private BufferedReader inFromServer;
	private ReaderThread readerThead;

	protected NetworkController() {
		try {
			webSocket = new Socket("127.0.0.1", 1337);
			outToServer = new PrintWriter(webSocket.getOutputStream(), true);
			inFromServer = new BufferedReader(new InputStreamReader(
					webSocket.getInputStream()));
			readerThead = new ReaderThread();
		} catch (Exception e) {
			// TODO make some retry for server connection
			e.printStackTrace();
		}
	}

	public static NetworkController getInstance() {
		if (instance == null) {
			instance = new NetworkController();
		}
		return instance;
	}

	public void init() {
		sendMessage("AIClientReady");
		readerThead.start();
	}

	public void readFunctionThread() {
		boolean exit = false;
		while (!exit) {
			String fullMessage = readMessage();

			if (fullMessage.equals("Net:OkForExit") || fullMessage == null) {
				exit = true;
				break;
			}

			dispatchMessage(fullMessage);
		}

	}

	public void dispatchMessage(String message) {
		String[] messageParts = message.split(":", 2);
		if (messageParts[0].equals("Net")) {
			parseNetMessage(messageParts[1]);
		} else if (messageParts[0].equals("Game")) {
			EventFactory.generateEvent(messageParts[1]);
		}
	}

	public void parseNetMessage(String message) {
		if (message.equals("JoinGameFailed")) {
			System.out.println("Error : the game client was not connected");
			System.out.println("Info : connection retry in few seconds...");
			try {
				Thread.sleep(2000);
				sendMessage("AIClientReady");
			} catch (InterruptedException e) {
				e.printStackTrace();
			}
		} else if (message.equals("ErrorGameClientDisconnect")) {
			System.out.println("Error : the game client was disconnected");
		} else {
			System.out.println("unknow net message: " + message);
		}
	}

	public void sendMessage(String message) {
		if (outToServer != null) {
			System.out.println(message);
			outToServer.println(message);
		} else {
			System.out.println("Error: can't send the message: " + message);
		}
	}

	public String readMessage() {
		String textServer = "";
		try {
			textServer = inFromServer.readLine();
			System.out.println("Server: " + textServer);
		} catch (IOException e) {
			e.printStackTrace();
		} catch (NullPointerException e) {
			System.out.println("Error: can't read on socket");
		}
		return textServer;
	}

	public void close() {
		try {
			sendMessage("Exit");
			readerThead.join();
			webSocket.close();
		} catch (IOException e) {
			e.printStackTrace();
		} catch (InterruptedException e) {
			e.printStackTrace();
		}
	}
}

File: AIClient_Python/src/network/NetworkController.py
'''
Created on Dec 17, 2014

@author: samuel
'''
import socket
import sys
import threading
import time
from event.EventFactory import EventFactory
from aiclient.Singleton import Singleton
from event.QueueController import QueueController

class NetworkController(object):
    _instance = None
    HOST, PORT = "localhost", 1337
    webSocket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    readerThread = None
    connected = False
    
    def __init__(self):
        self.readerThread = threading.Thread(target=self.readFunctionThread)
        try:
            self.webSocket.connect((self.HOST, self.PORT))
            self.connected = True
        except:
            print("Cannot connect to server", sys.exc_info()[0])

    def sendMessage(self, message):
        try:
            self.webSocket.sendall(message.encode())
        except:
            print("Error while sending to the socket", sys.exc_info()[0])
        
    def readMessage(self):
        try:
            message = self.webSocket.recv(1024)
            return message
        except:
            print("Error while reading from the socket", sys.exc_info()[0])
            return "".encode()
        
    def init(self):
        self.sendMessage("AIClientReady\n")
        self.readerThread.start()
    
    def executeOutgoingEvents(self):
        queueController = Singleton(QueueController)
        while not queueController.outEvents.empty():
            event = queueController.outEvents.get()
            self.sendMessage(event.toString() + "\n")
    
    def readFunctionThread(self):
        mustExit = False
        while not mustExit:
            message = self.readMessage().decode()
            
            if message == "Net:OkForExit\n" or message == "":
                mustExit = True
                self.connected = False
                break
            
            self.dispatchMessage(message)
    
    def dispatchMessage(self, message):
        messageParts = message.split(":", 1)
        if messageParts[0] == "Net":
            self.parseNetMessage(messageParts[1])
        elif messageParts[0] == "Game":
            EventFactory.generateEvent(messageParts[1])
        
    def parseNetMessage(self, message):
        if message == "JoinGameFailed":
            self.connectionRetry()
        elif message == "ErrorGameClientDisconnect":
            print("Error : the game client was disconnected")
        else:
            print("unknow net message:" + message)
    
    def closeConnection(self):
        self.sendMessage("Exit")
        self.readerThread.join()
        self.webSocket.close()
        self.connected = False
        
    def connectionRetry(self):
        print("Error : the game client was not connected")
        print("Info : connection retry in few seconds...")
        time.sleep(2)
        self.sendMessage("AIClientReady\n")